### PR TITLE
opt: use rocksdb merge operator to write single nodes

### PR DESCRIPTION
**Pull request by rphmeier**
The idea here is to use RocksDB [merge operators](https://github.com/facebook/rocksdb/wiki/Merge-Operator)
to avoid writing full pages when they are only slightly updated. In large commits, we update lots of pages, but many of them by only a tiny amount. However, the pages at the lowest levels of the trie are usually almost completely altered, so diffing is wasteful.

Whenever we update fewer than `FULL_PAGE_THRESHOLD=32` nodes in a page, we simply write a diff to RocksDB to be handled by the merge operator. When we update at least that many nodes, we opt to write the whole page. This parameter will require some tuning, but 32 gave good results on my end.

This eliminates most of the overhead of the WAL on pages reduces the time spent in `rocksdb_write` by about half again after #157 on my machine.

### CODE REVIEW

**Comment by rphmeier** [comment]
* **#153** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/153?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @rphmeier and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: nomt/src/store.rs [hunk 1]
```diff
@@ -130,3 +150,48 @@ impl Transaction {
...
+            let slot_end = slot_start + 32;
+            val[slot_start..slot_end].copy_from_slice(&op[1..]);
+        }
+    }
+
+    // merge operators are not supposed to fail unless there is corruption.
+    val
+}
+
+fn partial_merge_page_ops(
```
Comment (gabriele-0201) [inline nomt/src/store.rs:171]: I would add a comment here explaining that this is merging multiple merge operations into a new one. It's not obvious at first glance that `prev_value` is an older operation and `operands` are possibly multiple new operations

**Review by gabriele-0201 — APPROVED** [review]
Great improvement!

File: nomt/src/store.rs (hunk 1 above)
Comment (rphmeier) [reply nomt/src/store.rs:171]: yes, good point. I will link to the the RocksDB documentation that makes this clear.

**Pull request merged**